src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_pool_manager_core.clj

```diff
@@ -2,18 +2,24 @@
   (:require [schema.core :as schema]
             [puppetlabs.services.jruby-pool-manager.jruby-schemas :as jruby-schemas]
             [puppetlabs.services.jruby-pool-manager.impl.jruby-agents :as jruby-agents]
-            [puppetlabs.services.jruby-pool-manager.impl.jruby-internal :as jruby-internal]))
+            [puppetlabs.services.jruby-pool-manager.impl.jruby-internal :as jruby-internal])
+  (:import (puppetlabs.services.jruby_pool_manager.jruby_schemas ReferencePoolContext JRubyPoolContext)))
 
 (schema/defn ^:always-validate
   create-pool-context :- jruby-schemas/PoolContext
   "Creates a new JRuby pool context with an empty pool. Once the JRuby
   pool object has been created, it will need to be filled using `prime-pool!`."
   [config :- jruby-schemas/JRubyConfig]
-  (let [shutdown-on-error-fn (get-in config [:lifecycle :shutdown-on-error])]
-    {:config config
-     :internal {:modify-instance-agent (jruby-agents/pool-agent shutdown-on-error-fn)
-                :pool-state (atom (jruby-internal/create-pool-from-config config))
-                :event-callbacks (atom [])}}))
+  (let [shutdown-on-error-fn (get-in config [:lifecycle :shutdown-on-error])
+        pool-state (if (:multithreaded config)
+                     (atom (jruby-internal/create-reference-pool-from-config config))
+                     (atom (jruby-internal/create-pool-from-config config)))
+        pool-context-internal {:modify-instance-agent (jruby-agents/pool-agent shutdown-on-error-fn)
+                               :pool-state            pool-state
```

Comment:
i know this is controversial historically, but i think we agreed eventually that it's better not to align the second column like this